### PR TITLE
Accept generic unusual trades

### LINKED ISSUE #230: Generic Unusual Buy Orders
**Is your feature request related to a problem? Please describe.**
No

**Describe the feature you'd like**
Working generic unusual buy orders. Example: 
image

Have generic buy orders for unusuals accept any effect, whether it's priced or unpriced.

**Describe alternatives you've considered**

Only alternative solution is to constantly be aware of offer reviews from your bot and manually accept the trade sent, however this isn't automatic.

### PR DESCRIPTION
**Pull request by joekiller**
Allow bot to buy generic unusual orders.

Fixes #230

### CODE REVIEW

File: src/classes/Carts/UserCart.ts [hunk 1]
```diff
@@ -1937,7 +1936,7 @@ export default class UserCart extends Cart {
                 continue;
             }
 
-            const entry = this.bot.pricelist.getPrice(sku, true);
+            const entry = this.bot.pricelist.getPrice(sku, true, true);
```
[the remark this answers is not in the document]
Comment (joekiller) [reply src/classes/Carts/UserCart.ts]: fixed

File: src/classes/Carts/UserCart.ts [hunk 2]
```diff
@@ -1420,21 +1424,17 @@ export default class UserCart extends Cart {
...
                 } else {
-                    amount = amountCanTrade;
-                    alteredMessage = alteredMessage = `I can only sell ${amountCanTrade} more ${this.bot.schema.getName(
-                        SKU.fromString(sku),
-                        false
-                    )}`;
+                    alteredMessage = alteredMessage = `I can only sell ${skuCount.mostCanTrade} more ${skuCount.name}`;
 
-                    this.addOurItem(sku, amount);
+                    this.addOurItem(sku, skuCount.mostCanTrade);
```
[the remark this answers is not in the document]
Comment (joekiller) [reply src/classes/Carts/UserCart.ts:1433]: fixed

File: src/classes/Commands/functions/utils.ts [hunk 3]
```diff
@@ -56,22 +57,44 @@ export function getItemAndAmount(steamID: SteamID, message: string, bot: Bot): {
...
+        if (lowestUnusualDistance < lowestDistance) {
+            lowestDistance = lowestUnusualDistance;
+            closestMatch = closestUnusualMatch;
+            matchName = closestUnusualMatch.name;
+            // don't mutate any objects from getPrices
+            closestMatch = closestMatch.clone();
+            closestMatch.name = closestUnusualMatch.name.replace('Unusual', genericEffect.name);
+            closestMatch.sku = `${closestUnusualMatch.sku};u${genericEffect.effect.id}`;
+        }
+
```
[the remark this answers is not in the document]
Comment (joekiller) [reply src/classes/Commands/functions/utils.ts]: added more details

File: src/classes/Carts/Cart.ts [hunk 4]
```diff
@@ -126,10 +126,34 @@ export default abstract class Cart {
         return this.our[sku] !== undefined ? this.our[sku].amount : 0;
     }
 
+    getOurGenericCount(sku: string): number {
+        return this.getGenericCount(sku, s => this.getOurCount(s));
+    }
+
+    getGenericCount(sku: string, getCountFn: (sku: string) => number): number {
+        const pSku = SKU.fromString(sku);
+        if (pSku.quality === 5) {
```
Comment (Bonfire) [inline src/classes/Carts/Cart.ts:135]: Do we not have an enum for all item qualities? If not, possible addition for the future? (Would result in us able to use something like `QUALITY.UNUSUAL`)

File: src/classes/Carts/Cart.ts [hunk 5]
```diff
@@ -126,10 +126,34 @@ export default abstract class Cart {
...
+            // try to count all unusual types
+            return this.bot.unusualEffects
+                .map(e => {
+                    pSku.effect = e.id;
+                    const s = SKU.fromObject(pSku);
+                    return getCountFn(s);
+                })
+                .reduce((p, c) => (p ? p + c : c));
```
Comment (Bonfire) [inline src/classes/Carts/Cart.ts:143]: Are we able to make these more descriptive for future additions?

File: src/classes/Carts/UserCart.ts [hunk 6]
```diff
@@ -431,19 +431,15 @@ export default class UserCart extends Cart {
...
+                if (skuCount.mostCanTrade === 0) {
+                    alteredMessage = `I can't sell more ${skuCount.name}`;
                     this.bot.listings.checkBySKU(sku);
                 } else {
-                    amount = amountCanTrade;
-                    alteredMessage = `I can only sell ${amountCanTrade} more ${this.bot.schema.getName(
-                        SKU.fromString(sku),
-                        false
-                    )}`;
+                    alteredMessage = `I can only sell ${skuCount.mostCanTrade} more ${skuCount.name}`;
```
Comment (Bonfire) [inline src/classes/Carts/UserCart.ts:442]: Does `skuCount.name` here require optional pluralization?

File: src/classes/Inventory.ts [hunk 7]
```diff
@@ -195,6 +204,21 @@ export default class Inventory {
...
+    getAmountOfGenerics(sku: string, tradableOnly?: boolean): number {
+        const s = SKU.fromString(sku);
+        if (s.quality === 5) {
+            // generic getAmount so return total that match the generic sku type
+            return this.unusualEffects
+                .map(e => {
+                    s.effect = e.id;
+                    return this.getAmount(SKU.fromObject(s), tradableOnly);
+                })
+                .reduce((p, c) => (p ? p + c : c));
```
Comment (Bonfire) [inline src/classes/Inventory.ts:216]: Same comment as above (potential refactoring of variable names here for clarity)

**Review by Bonfire — APPROVED** [review]
Changes look great. Any comments I've made are "non-showstoppers". Thank you for the PR, Joe

File: src/classes/Carts/Cart.ts (hunk 4 above)
Comment (joekiller) [reply src/classes/Carts/Cart.ts:135]: we do not have an enum for all the qualities. I just use what's in the schema and figure that is more future proof as the schema gets updated.

File: src/classes/Carts/Cart.ts (hunk 5 above)
Comment (joekiller) [reply src/classes/Carts/Cart.ts]: I updated the code

File: src/classes/Inventory.ts (hunk 7 above)
Comment (joekiller) [reply src/classes/Inventory.ts]: Updated

File: src/classes/Carts/UserCart.ts (hunk 6 above)
Comment (joekiller) [reply src/classes/Carts/UserCart.ts]: Corrected

**Comment by Bonfire** [comment]
Updating approval for most recent changes

**Review by Bonfire — APPROVED** [review]

**Review by idinium96 — APPROVED** [review]
I will merge this soon, or you can do it too! :)

**Pull request merged**